Detectors/EMCAL/simulation/src/Detector.cxx

```diff
@@ -135,101 +155,145 @@ void Detector::ConstructGeometry()
-              << fMC->CurrentVolOffName(3);
-
-  // porting from aliroot: circumvention of the index problem. GetAbsCellID cannot be used
-  // Int_t detID = geom->GetAbsCellId(offset + copySmod - 1, copyMod - 1, copyPhi - 1, copyEta - 1);
-  //
-  // Swap A side in Phi and C side in Eta due to wrong indexing.
-  //
-  Int_t smNumber = offset + copySmod - 1, smTypeID = 1;
-  auto [iphi, ieta] = geom->GetCellPhiEtaIndexInSModule(smNumber, copyMod - 1, copyPhi - 1, copyEta - 1);
-  //iphi = std::get<0>(posetaphi);
-  //ieta = std::get<1>(posetaphi);
-  if (smNumber % 2 == 0) {
-    if (strcmp(fMC->CurrentVolOffName(4), "DCSM") == 0)
-      smTypeID = 3; //DCal supermodule. previous design/idea
-    else
-      smTypeID = 2;
-    ieta = ((geom->GetCentersOfCellsEtaDir()).size() * 2 / smTypeID - 1) - ieta; // 47/31-ieta, revert the ordering on A side in order to keep convention.
-  } else {
-    if (strcmp(fMC->CurrentVolOffName(4), "SM10") == 0)
-      smTypeID = 2; //half supermodule. previous design/idea
-    if (strcmp(fMC->CurrentVolOffName(4), "SM3rd") == 0)
-      smTypeID = 3; //one third (installed in 2012) supermodule
-    if (strcmp(fMC->CurrentVolOffName(4), "DCEXT") == 0)
-      smTypeID = 3;                                                          //one third (installed in 2012) supermodule
-    iphi = ((geom->GetCentersOfCellsPhiDir()).size() / smTypeID - 1) - iphi; // 23/7-iphi, revert the ordering on C side in order to keep convention.
+  auto track = fMC->GetStack()->GetCurrentTrackNumber(),
+       directparent = fMC->GetStack()->GetCurrentParentTrackNumber();
+  if (track != mCurrentTrack) {
+    LOG(DEBUG4) << "Doing new track " << track << " current (" << mCurrentTrack << "), direct parent (" << directparent << ")" << std::endl;
+    // new current track - check parentage
+    auto hasSuperParent = mSuperParentsIndices.find(directparent);
+    if (hasSuperParent != mSuperParentsIndices.end()) {
+      // same superparent as direct parent
+      mCurrentParentID = hasSuperParent->second;
+      mSuperParentsIndices[track] = hasSuperParent->second;
+      auto superparent = mSuperParents.find(mCurrentParentID);
+      if (superparent != mSuperParents.end()) {
+        mCurrentSuperparent = &(superparent->second);
+      } else {
+        LOG(ERROR) << "Attention: No superparent object found (parent " << mCurrentParentID << ")";
+        mCurrentSuperparent = nullptr;
+      }
+      LOG(DEBUG4) << "Found superparent " << mCurrentParentID << std::endl;
+    } else {
+      // start of new chain
+      // for new incoming tracks the super parent index is equal to the track ID (for recursion)
+      mSuperParentsIndices[track] = track;
+      mCurrentSuperparent = AddSuperparent(track, fMC->TrackPid(), fMC->Etot());
+      mCurrentParentID = track;
+    }
+    mCurrentTrack = track;
   }
+  if (std::string_view(v->GetName()) == "SCMX") {
```

Comment:
this string comparison can be avoided by simply caching the volID of SCMX

Reply:
Done. Now all string comparisons are removed from ProcessHits.